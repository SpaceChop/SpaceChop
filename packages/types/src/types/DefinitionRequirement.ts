import * as t from 'runtypes';

const DefinitionRequirement = t.Union(
  t.Literal('faces'),
);

export type DefinitionRequirement = t.Static<typeof DefinitionRequirement>;

export default DefinitionRequirement;
