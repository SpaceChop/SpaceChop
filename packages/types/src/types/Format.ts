import * as t from 'runtypes';

const Format = t.Union(
  t.Literal('jpeg'),
  t.Literal('png'),
  t.Literal('gif'),
  t.Literal('webp'),
);

const Mime = t.Union(
  t.Literal('image/jpeg'),
  t.Literal('image/png'),
  t.Literal('image/gif'),
  t.Literal('image/webp'),
);

export type Mime = t.Static<typeof Mime>;
export type Format = t.Static<typeof Format>;

export const allFormats = Format.alternatives.map((f) => f.value);
export const allMimes = Mime.alternatives.map((f) => f.value);

export const formatToMime = (format: Format): Mime => {
  const typedFormat = Format.check(format);
  const mime = ('image/' + typedFormat) as Mime;
  return mime;
};

export const mimeToFormat = (mime: Mime): Format => {
  const typedMime = Mime.check(mime);
  const type = typedMime.match(/^image\/(\w+)$/)[1] as Format;
  return type;
};

export const parseFormat = (value: string): Format => {
  switch (value) {
    case 'jpg':
      return 'jpeg' as Format;
    default:
      return value in allFormats ? value as Format : null;
  }
};

export default Format;
