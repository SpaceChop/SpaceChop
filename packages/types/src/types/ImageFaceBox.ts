import * as t from 'runtypes';

const ImageFaceBox = t.Record({
  x: t.Number,
  y: t.Number,
  width: t.Number,
  height: t.Number,
});

export type ImageFaceBox = t.Static<typeof ImageFaceBox>;
export default ImageFaceBox;
