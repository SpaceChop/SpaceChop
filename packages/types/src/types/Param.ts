import * as t from 'runtypes';

const Param = t.Record({
  from_path: t.String,
});

export type Param = t.Static<typeof Param>;
export default Param;
