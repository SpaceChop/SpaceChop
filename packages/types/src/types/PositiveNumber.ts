import * as t from 'runtypes';

const PostiveNumber = t.Number.withConstraint((n) => {
  if (n < 0) {
    return 'please no negative number';
  }
  return true;
});

export type Gravity = t.Static<typeof PostiveNumber>;
export default PostiveNumber;
