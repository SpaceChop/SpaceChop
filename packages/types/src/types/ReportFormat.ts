import * as t from 'runtypes';

const Format = t.Union(
  t.Literal('json'),
);

export const allFormats = Format.alternatives.map((f) => f.value);
export type Format = t.Static<typeof Format>;
export default Format;
