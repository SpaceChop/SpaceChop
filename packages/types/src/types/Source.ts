import * as t from 'runtypes';
import HttpSourceConfig from './sources/http';
import S3SourceConfig from './sources/s3';
import VolumeSourceConfig from './sources/volume';

const Source = t.Union(
  t.Record({ http: HttpSourceConfig }),
  t.Record({ s3: S3SourceConfig }),
  t.Record({ volume: VolumeSourceConfig }),
);

export type Source = t.Static<typeof Source>;
export default Source;
