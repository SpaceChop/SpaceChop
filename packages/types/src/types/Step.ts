import * as t from 'runtypes';
import CompressConfig from './operations/compress';
import CropConfig from './operations/crop';
import FillConfig from './operations/fill';
import FitConfig from './operations/fit';
import FormatConfig from './operations/format';
import ResizeConfig from './operations/resize';
import StripConfig from './operations/strip';

const Step = t.Partial({
  $compress: CompressConfig,
  $crop: CropConfig,
  $fill: FillConfig,
  $fit: FitConfig,
  $format: FormatConfig,
  $resize: ResizeConfig,
  $strip: StripConfig,
}).withConstraint(
  (n) => {
    if (Object.keys(n).length > 1) {
      return 'please only one operation per step';
    }
    if (Object.keys(n).length === 0) {
      return 'please no empty step';
    }
    return true;
  },
);

export type Step = t.Static<typeof Step>;
export default Step;
