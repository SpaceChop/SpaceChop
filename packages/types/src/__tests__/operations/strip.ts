import { defaultConfig } from '.';
import { Config } from '../../types/Config';
import validate from '../../validate';

describe('should validate strip config', () => {
  describe('should throw on invalid config', () => {
    it('require valid icc_profile', () => {
      const config = {
        ...defaultConfig,
        presets: {
          t_200: {
            steps: [{
              $strip: {
                icc_profile: 'icc_profile',
              },
            }],
          },
        },
      };
      expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('should now throw on valid config', () => {
    it(`valid icc_profile`, () => {
      const config: Config = {
        ...defaultConfig,
        presets: {
          t_200: {
            steps: [{
              $strip: {
                icc_profile: true,
              },
            }],
          },
        },
      };
      expect(() => validate(config)).not.toThrow();
    });
  });
});
