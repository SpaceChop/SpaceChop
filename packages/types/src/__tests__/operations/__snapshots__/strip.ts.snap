// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should validate operations no double operations in step 1`] = `"presets.t_200.steps.[0]: please only one operation per step"`;

exports[`should validate strip config should throw on invalid config require valid icc_profile 1`] = `"presets.t_200.steps.[0].$strip.icc_profile: Expected boolean, but was string"`;
